### PR TITLE
tests/drivers/ipm: Support ipm test case in SMP

### LINKED ISSUE #12478: tests/drivers/ipm/peripheral.mailbox failing sporadically on qemu_x86_64 (timeout)
see https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/31042/6/tests, the test times out with

```
***** Booting Zephyr OS zephyr-v1.13.0-3219-gdfc8f0a6d5 *****
starting test - Test IPM
ipm_recv0: 'everything is awesome'
ipm_recv0: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor '

```

### PR DESCRIPTION
**Pull request by wentongwu**
tests/drivers/ipm: support ipm test case in SMP
    
    Current design of ipm test doesn't consider SMP, there is the
    possibility that offload_irq and rx_thread run at different
    cpu core at the same time, it will cause state(enabled and
    channel_disable) disorder if two different context access them concurrently.
    
    Fixes #12478.
    
    Signed-off-by: Wentong Wu <wentong.wu@intel.com>

### CODE REVIEW

File: drivers/console/ipm_console_receiver.c [hunk 1]
```diff
@@ -66,24 +66,12 @@ static void ipm_console_thread(void *arg1, void *arg2, void *arg3)
...
-			key = irq_lock();
-			ipm_set_enabled(driver_data->ipm_device, 1);
-			driver_data->channel_disabled = 0;
-			irq_unlock(key);
-		}
 	}
 }
 
+#ifdef CONFIG_SMP
+__attribute__((optimize("O0")))
```
Comment (andrewboie) [inline drivers/console/ipm_console_receiver.c]: Ouch. Why do we have to do this?

File: tests/drivers/ipm/src/main.c [hunk 2]
```diff
@@ -102,6 +104,9 @@ void main(void)
 	 * automation purposes?
 	 */
 
+#ifdef CONFIG_SMP
+	k_sleep(1000 * DELAY_US);
```
Comment (andrewboie) [inline tests/drivers/ipm/src/main.c]: What is this sleep for?

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
In addition to my comments, the git history needs to be cleaned up, we have a merge commit and 2 patches with the same commit message here. You need to interactively rebase.

File: tests/drivers/ipm/src/main.c (hunk 2 above)
Comment (andyross) [inline tests/drivers/ipm/src/main.c]: Why not DELAY_MS then?

Also: why?

File: tests/drivers/ipm/testcase.yaml [hunk 3]
```diff
@@ -2,5 +2,5 @@ tests:
   peripheral.mailbox:
     filter: not CONFIG_SOC_QUARK_SE_C1000_SS
     arch_exclude: posix xtensa
-    platform_exclude: qemu_x86_64 # see issue #12478
+    #platform_exclude: qemu_x86_64 # see issue #12478
```
Comment (andyross) [inline tests/drivers/ipm/testcase.yaml]: Please remove this line if you're fixing the reason for the exclude, don't comment it out.

**Review by andyross — CHANGES_REQUESTED** [review]
A few nitpicks.  Broadly though, I don't understand why this fixes what it says it fixes.  How exactly was this broken in SMP?  The commit message should have an explanation.

File: drivers/console/ipm_console_receiver.c (hunk 1 above)
Comment (wentongwu) [reply drivers/console/ipm_console_receiver.c]: without this, compiler will do some optimization for “while (ring_buf_space_get(&driver_data->rb) == 0）{}； ”  and it only call "ring_buf_space_get(&driver_data->rb) " once. And it broken monitoring ring buffer's space.

File: tests/drivers/ipm/src/main.c (hunk 2 above)
Comment (wentongwu) [reply tests/drivers/ipm/src/main.c]: For SMP, maybe console rx_thread haven't done the printout, but main thread will complete it's work, so add some delay to wait rx thread.

File: drivers/console/ipm_console_receiver.c (hunk 1 above)
Comment (andyross) [reply drivers/console/ipm_console_receiver.c]: If that's true, then the problem is a missing "volatile" somewhere underneath the inlined implementation of ring_buf_space_get().  We should only be changing optimizer flags for known optimizer bugs.

File: tests/drivers/ipm/testcase.yaml (hunk 3 above)
Comment (wentongwu) [reply tests/drivers/ipm/testcase.yaml]: yes, I will remove it.

File: tests/drivers/ipm/src/main.c (hunk 2 above)
Comment (wentongwu) [reply tests/drivers/ipm/src/main.c]: Ok , I will use DELAY_MS. Thanks

File: drivers/console/ipm_console_receiver.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (wentongwu) [reply drivers/console/ipm_console_receiver.c]: for  "Broadly though, I don't understand why this fixes what it says it fixes. How exactly was this broken in SMP? The commit message should have an explanation."

And prior this PR, with SMP on, there is the possibility that offload_irq and rx_thread run at different cpu core at the same time, it will access ipm_set_enable and channel_disable at the same time and causing state(enabled and channel_disable) wrong. And this PR just use two threads(main thread and rx thread) and sync up each other with sem and ring buffer's space. It will never access same thing at the same time.

File: tests/drivers/ipm/src/main.c (hunk 2 above)
Comment (andrewboie) [reply tests/drivers/ipm/src/main.c]: I'd feel a lot better if this was synchronized with a semaphore instead.

File: drivers/console/ipm_console_receiver.c (hunk 1 above)
Comment (wentongwu) [reply drivers/console/ipm_console_receiver.c]: yes, will add volatile for ring_buffer's head tail and size.

File: tests/drivers/ipm/src/main.c (hunk 2 above)
Comment (wentongwu) [reply tests/drivers/ipm/src/main.c]: yes, but for now ipm structure doesn't provide mechanism to notify printout done. It's not elegant to just add a semphore or atomic variable in two files. And I would like to review code more to do it. And also I run the test about 1000 times with current design, no failure happened.

**Comment by wentongwu** [comment]
@andyross  @andrewboie could you please help review the new code? thanks

File: include/ring_buffer.h [hunk 4]
```diff
@@ -25,8 +25,8 @@ extern "C" {
  * @brief A structure to represent a ring buffer
  */
 struct ring_buf {
-	u32_t head;	 /**< Index in buf for the head element */
-	u32_t tail;	 /**< Index in buf for the tail element */
+	volatile u32_t head;	 /**< Index in buf for the head element */
+	volatile u32_t tail;	 /**< Index in buf for the tail element */
```
Comment (andyross) [inline include/ring_buffer.h]: I think this is the wrong spot.   There are plenty of uses for a ring buffer that don't involve contending with SMP CPUs.  Instead, I think you probably want to tag the pointer to the struct volatile in the app instead.

File: include/spinlock.h [hunk 5]
```diff
@@ -33,7 +33,7 @@ static inline void z_arch_irq_unlock(int key)
  * even more poorly supported.
  */
 #if (CONFIG_FLASH_SIZE == 0) || (CONFIG_FLASH_SIZE > 32)
-#if defined(CONFIG_ASSERT) && (CONFIG_MP_NUM_CPUS < 4)
+#if defined(CONFIG_ASSERT) && (CONFIG_MP_NUM_CPUS < 2)
```
Comment (andyross) [inline include/spinlock.h]: No... that's wrong.  This disables spinlock validation whenever we're running with more than one CPU.  Why is this change here?  The test for CPU count is because I'm (somewhat sneakily) packing the CPU index into the bottom two bits of the current thread pointer for testing ownership of the spinlock, and there are only two such bits available.

I know @dcpleung reported some problems with validation in SMP on hardware too (though oddly I don't see them on x86_64), so if you need to disable SPIN_VALIDATE on SMP that's probably OK.  Just do it in a separate test and not in this line.

**Review by andyross — CHANGES_REQUESTED** [review]
One nitpick about volatile, one spot in the SPIN_VALIDATE predication that definitely needs to be changed.

File: include/spinlock.h (hunk 5 above)
Comment (wentongwu) [reply include/spinlock.h]: @andyross  thanks for the review. 
While I'm doing the debug for this case, I found when two different threads running on different cpu try to access the same one spinlock or semaphore concurrently, there will be assert. The situation is that one thread get the lock(set the owner) and didn't do the unlock, the other thread exactly do the spin validate.

OK, I will do the disable in another PR.

File: include/ring_buffer.h (hunk 4 above)
Comment (wentongwu) [reply include/ring_buffer.h]: Ok, Thanks

Comment (wentongwu) [reply include/ring_buffer.h]: > One nitpick about volatile, one spot in the SPIN_VALIDATE predication that definitely needs to be changed.
I have done the code change, but I should git push it after #14863 has been merged, otherwise it cann't pass the shippable. So please review #14863 first! Thanks!

Comment (gnuless) [reply include/ring_buffer.h]: Hate to stick my nose in here, but is `volatile` even the right thing here?  If the issue is that other CPUs might change the value, then I'm not convinced that we don't also need explicit storage fences.  Probably not a problem on x86, but on other arches.

Comment (andyross) [reply include/ring_buffer.h]: On architectures where it matters, the fence will be provided by the synchronization primitive where we are using one, which we aren't here.  It's true that IPM is sort of a special case, as it was intended for a asymmetric multiprocessing architecture (Quark SE -- one x86 and one ARC on the same memory bus) with no cache.  But that's not really something we can fix with this patch.

**Comment by wentongwu** [comment]
@andyross could you please review the new changes?

File: drivers/console/ipm_console_receiver.c [hunk 6]
```diff
@@ -66,51 +69,29 @@ static void ipm_console_thread(void *arg1, void *arg2, void *arg3)
 			++pos;
 		}
 
-		/* ISR may have disabled the channel due to full buffer at
-		 * some point. If that happened and there is now room,
-		 * re-enable it.
```
Comment (andyross) [inline drivers/console/ipm_console_receiver.c]: I don't understand why it's OK to remove this.  That condition (an interrupt delivering data that fills the buffer) seems like it can still happen.

File: drivers/console/ipm_console_receiver.c [hunk 7]
```diff
@@ -66,51 +69,29 @@ static void ipm_console_thread(void *arg1, void *arg2, void *arg3)
...
-		}
 	}
 }
 
 static void ipm_console_receive_callback(void *context, u32_t id,
 					 volatile void *data)
 {
 	struct device *d;
-	struct ipm_console_receiver_runtime_data *driver_data;
+	struct ipm_console_receiver_runtime_data *volatile driver_data;
```
Comment (andyross) [inline drivers/console/ipm_console_receiver.c]: The volatile is incorrect here.  This delcares a volatile **pointer** (i.e. it tells the compiler that the memory where the local driver_data variable is stored may change and must not be optimized into a register, which doesn't make sense because no other thread is going to be mutating our stack like that).  You want a pointer to a volatile struct instead:

    volatile struct ipm_console_receiver_runtime_data *driver_data;

File: drivers/console/ipm_console_receiver.c [hunk 8]
```diff
@@ -66,51 +69,29 @@ static void ipm_console_thread(void *arg1, void *arg2, void *arg3)
...
+	key = k_spin_lock(&driver_data->rb_spinlock);
 	/* Should always be at least one free buffer slot */
 	ret = ring_buf_item_put(&driver_data->rb, 0, id, NULL, 0);
+	k_spin_unlock(&driver_data->rb_spinlock, key);
 	__ASSERT(ret == 0, "Failed to insert data into ring buffer");
 	k_sem_give(&driver_data->sem);
 
-	/* If the buffer is now full, disable future interrupts for this channel
-	 * until the thread has a chance to consume characters.
-	 *
```
Comment (andyross) [inline drivers/console/ipm_console_receiver.c]: Same question as before: why is it OK to remove this?  The condition and the comment explaining it (a RX interrupt is out of space in the buffer) makes sense, and I don't see anything in the patch that addresses it.

File: tests/drivers/ipm/src/ipm_dummy.c [hunk 9]
```diff
@@ -69,13 +69,8 @@ static int ipm_dummy_send(struct device *d, int wait, u32_t id,
 	driver_data->regs.id = id;
 	driver_data->regs.busy = 1U;
 
-	irq_offload(ipm_dummy_isr, d);
+	ipm_dummy_isr(d);
 
-	if (wait) {
-		while (driver_data->regs.busy) {
-			/* busy-wait */
```
Comment (andyross) [inline tests/drivers/ipm/src/ipm_dummy.c]: Why is this removed?  It sounds like the original test expects the busy flag to be cleared by an ISR before returning, and now it's going to return synchronously before the interrupt (or whatever) reports success.

**Review by andyross — CHANGES_REQUESTED** [review]
First, the fact that this is hanging on x86_64 in the CI test implies that this probably doesn't fix the bug it claims to fix. :)

Beyond that, the meat of the changes seem to make sense: you need to have locking around the ring buffer in SMP where you don't in UP, and you need to add a sleep in the test because the other CPU may try to mutate the thing before your test code is prepared for the next step.  Those seem fine.

But I don't understand the removal of the (fairly clearly explained) buffer full handling at all.  Why is this not introducing a failure case when the buffer fills up?

File: drivers/console/ipm_console_receiver.c (hunk 7 above)
Comment (wentongwu) [reply drivers/console/ipm_console_receiver.c]: Yes, you are right, I didn't get the right usage of volatile in that place, and above change should be the cause that make the CI test fail.

File: drivers/console/ipm_console_receiver.c (hunk 8 above)
Comment (wentongwu) [reply drivers/console/ipm_console_receiver.c]: sender thread will wait until there is space in ring buffer to do next send.
+       while (ring_buf_space_get(&driver_data->rb) == 0) {
+       }

File: tests/drivers/ipm/src/ipm_dummy.c (hunk 9 above)
Comment (wentongwu) [reply tests/drivers/ipm/src/ipm_dummy.c]: as above, sender thread will wait until there is at least one free buffer for the ring buffer.

File: drivers/console/ipm_console_receiver.c (hunk 6 above)
Comment (wentongwu) [reply drivers/console/ipm_console_receiver.c]: I don't want to use the flags(enabled and channel_disable), there is possibility that the flags are accessed by ISR and receiver thread at the same time if SMP enabled.

**Comment by andyross** [comment]
OK, I see the spinning behavior and sort of understand.  And to be fair, IPM is a mostly unused, oddball subsystem.  But... that's a fairly major API change!  You're changing the behavior of the code from "disable and return a synchronous error" to "spin forever if the other side doesn't read".  That's not really in the spirit of "bug fix for SMP".

And I don't understand why you have to do it.  The same tricks that work with this code as written (it was originally for a multiprocessing setup!) should be able to work when there are extra cores on one side, with a little locking.

**Review by andyross — CHANGES_REQUESTED** [review]
FWIW, I took a look today.  It really does seem like this is only an SMP safety thing, and we've been working around that by disabling SMP for 1.14.  We can come back to this in 1.15 and do it right.
